File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
typedef pair<int, int> pii;
typedef pair<int, pii> pipii;

class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        // Making the adjacency graph
        vector<vector<pii>> graph(n);

        for(const auto &flight: flights){
            int from = flight[0]; 
            int to = flight[1]; 
            int price = flight[2];

            graph[from].push_back({to, price}); 
        }

        int cost[n];
        fill_n(cost, n, INT_MAX);
        cost[src] = 0;

        // Min heap: {cost, currentNode, moves}
        priority_queue<pipii, vector<pipii>, greater<pipii>> pq;
        pq.push({0, {src, 0}});

        while(!pq.empty()){
            int currentCost = pq.top().first;
            int current = pq.top().second.first;
            int moves = pq.top().second.second;
            pq.pop();

            if(current == dst) return currentCost;
            
            if(moves <= k){
                for(const auto &next : graph[current]){
                    int to = next.first;
                    int price = next.second;

                    int newCost = currentCost + price;
                    if(newCost < cost[to]){
                        cost[to] = newCost;
                        pq.push({newCost, {to, moves + 1}});
                    }
                }
            }
        }

        return -1;
    }
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        List<List<int[]>> adj = new ArrayList<>(); // node -> (adj, cost), (adj, cost)
        for(int i = 0; i < n; i++){
            adj.add(new ArrayList<>());
        }

        for(int[] flight : flights){
            int source = flight[0], dest = flight[1], cost = flight[2];

            adj.get(source).add(new int[]{dest, cost});
        }


        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));

        pq.offer(new int[]{0, 0, src}); // {stops, distance, node}

        int[] distance = new int[n];
        Arrays.fill(distance, Integer.MAX_VALUE);

        while(!pq.isEmpty()){
            int[] front = pq.poll();

            int stops = front[0], d = front[1], node = front[2];

            if(stops <= k){
                for(int[] edge : adj.get(node)){
                    int adjNode = edge[0], cost = edge[1];

                    if(d + cost < distance[adjNode]){
                        distance[adjNode] = d + cost;
                        pq.offer(new int[]{stops+1, distance[adjNode], adjNode});
                    }
                }
            }
        }

        return distance[dst] == Integer.MAX_VALUE ? -1 : distance[dst];
    }
}